edx_exams/apps/core/management/commands/bulk_add_course_staff.py

```diff
@@ -0,0 +1,98 @@
+import time
+
+import unicodecsv
+from django.core.management.base import BaseCommand
+from django.db import transaction
+
+from edx_exams.apps.core.models import CourseStaffRole, User
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    Management command to add Course Staff (and User) in batches from CSV
+    """
+    help = """
+    Add Course Staff in bulk from CSV.
+    Expects that the data will be provided in a csv file format with the first row
+    being the header and columns being: username, email, role, course_id.
+    Example:
+        $ ... bulk_add_course_staff --csv_path=foo.csv
+    """
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-p', '--csv_path',
+            metavar='csv_path',
+            dest='csv_path',
+            required=False,
+            help='Path to CSV file.')
+        parser.add_argument(
+            '--batch_size',
+            type=int,
+            default=200,
+            dest='batch_size',
+            help='Batch size')
+        parser.add_argument(
+            '--batch_delay',
+            type=float,
+            default=1.0,
+            dest='batch_delay',
+            help='Time delay in each iteration')
+
+    @transaction.atomic
+    def handle(self, *args, **options):
+        """
+        The main logic and entry point of the management command
+        """
+        csv_path = options['csv_path']
+        batch_size = options['batch_size']
+        batch_delay = options['batch_delay']
+
+        if csv_path:
+            with open(csv_path, 'rb') as csv_file:
+                self.add_course_staff_from_csv(csv_file, batch_size, batch_delay)
+
+        logger.info('Bulk add course staff complete!')
+
+    def add_course_staff_from_csv(self, csv_file, batch_size, batch_delay):
+        """
+        Add the given set of course staff provided in csv
+        """
+        reader = list(unicodecsv.DictReader(csv_file))
+        users_to_create = []
+        users_existing = {u.username for u in User.objects.filter(username__in=[r.get('username') for r in reader])}
+        for row in reader:
+            # if not User.objects.filter(username=row.get('username')).exists():
+            #     users_to_create.append(row)
+            if row.get('username') not in users_existing:
+                users_to_create.append(row)
+
+        # bulk create users
+        for i in range(0, len(users_to_create), batch_size):
+            User.objects.bulk_create(
+                User(
+                    username=user.get('username'),
+                    email=user.get('email'),
+                )
+                for user in users_to_create[i:i + batch_size]
+            )
+            time.sleep(batch_delay)
+
+        # bulk create course staff
+        for i in range(0, len(reader), batch_size):
+            CourseStaffRole.objects.bulk_create(
+                CourseStaffRole(
+                    user=User.objects.get(
```

Comment:
I'm not sure about the answer to this question, and maybe this is not possible. But I noticed that you create the `users_existing` set querying the `User` model via the `username` field. Here, though, you're querying the `User` model via the `username` _and_ the `email` field. I'm wondering whether a scenario is possible where a user's would be found in `users_existing` but not in this query.

If we're pulling this data from the LMS, I think this could happen, because `email` is modifiable, but `username` isn't. I think `email` is maintained on our `User` model via syncing with the `JWT` when they hit `edx-exams` in the process of an exam, right? Do you think this is something we need to address here? I'm just not sure this code guarantees a user will exist, which could cause an exception here.

Since the function is in a transaction, it should be okay, because it'll rollback, we can correct the CSV, and try again, but it would be better to handle explicitly.

Reply:
Oooh, that's a really good point. Since the username is unique, could I just get by username instead? I've updated to be that way but let me know if that could cause any issues.

Reply:
I think that's good! `username` should not change.